test/main.test.js

```diff
@@ -129,4 +129,40 @@ describe('LoopBack Context', function() {
       ], done);
     });
   });
+  it('handles concurrent then() calls with when v3.7.7 promises & bind option',
+  function() {
+    var timeout = 25;
+    function runWithPushedValue(pushedValue, bind, expectToFail) {
+      return new Promise(function concurrentExecution(outerResolve, reject) {
+        LoopBackContext.runInContext(function pushToContext() {
+          var ctx = LoopBackContext.getCurrentContext({bind: bind});
+          expect(ctx).is.an('object');
+          ctx.set('test-key', pushedValue);
+          var whenPromise = whenV377.promise(function(resolve) {
+            setTimeout(resolve, timeout);
+          });
+          whenPromise.then(function pullFromContext() {
+            var pulledValue = ctx && ctx.get('test-key');
+            return pulledValue;
+          }).then(function verify(pulledValue) {
+            if (bind) {
+              expect(pulledValue).to.equal(pushedValue);
+            } else if (expectToFail) {
+              expect(pulledValue).to.not.equal(pushedValue);
+            }
+            outerResolve();
+          }).catch(reject);
+        });
+      });
+    }
+    return Promise.all([
+      runWithPushedValue('test-value-1', true),
+      runWithPushedValue('test-value-2', true),
+    ]).then(function() {
+      return Promise.all([
+        runWithPushedValue('test-value-3'),
+        runWithPushedValue('test-value-4', false, true),
```

Comment:
I think the race condition is caused by the fact that sometimes the first runner fails while other times it's the second runner that fails.

I think we need to rework the tests in such way that the runner only report the value observed in the context, and then the tests asserts that either 1) all values are correct 2) one of the values is not correct, but it doesn't matter which one.

```js
// please use more descriptive test names than "works" and "does not work" :)
it('works', () => {
  return Promise.all([
    runWithPushedValue('test-value-1', {bind: true}),
    runWithPushedValue('test-value-2', {bind: true}),
  ]).then((results) => {
    expect(result).to.eql(['test-value-1', 'test-value-2']);
  });
});

it('does not work', () => {
  return Promise.all([
    runWithPushedValue('test-value-1', {bind: false}),
    runWithPushedValue('test-value-2', {bind: false}),
  ]).then((results) => {
    expect(result).to.not.eql(['test-value-1', 'test-value-2']);
  });
});

function runWithPushedValue(...) {
  // ...
  whenPromise.then(function pullFromContext() {
    var pulledValue = ctx && ctx.get('test-key');
    outerResolve(pulledValue);
  })
  .catch(reject);
}
```